### PR TITLE
hindsight: init at 2021.01.16

### PR DESCRIPTION
**Pull request by ryneeverett**
###### Motivation for this change

Web browser forensics for Google Chrome/Chromium

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/node-packages/composition.nix
Comment (SuperSandro2000) [inline pkgs/development/node-packages/composition.nix:9]: Please revert all those changes. Also node2nix is currently broken and does not eval on master.

File: pkgs/development/python-modules/puremagic/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/puremagic/default.nix:3]:
```suggestion
{ lib, fetchFromGitHub }:
```

Please inputs python packages directly and add ``disabled = isPy27;``.

File: pkgs/development/python-modules/puremagic/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/puremagic/default.nix:30]:
```suggestion
```

set by buildPythonApplication

File: pkgs/development/python-modules/xlsx2html/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/xlsx2html/default.nix:1]:
```suggestion
{ lib, pythonPackages, fetchPypi }:
```
This breaks the package on python 3.7, 3.9 and 3.10. ``disabled = isPy27;``

File: pkgs/development/python-modules/xlsx2html/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/xlsx2html/default.nix:3]:
```suggestion
buildPythonPackage rec {
```

File: pkgs/development/python-modules/xlsx2html/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/xlsx2html/default.nix:12]:
```suggestion
  propagatedBuildInputs = [
```

File: pkgs/development/tools/hindsight/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/hindsight/default.nix:4]:
```suggestion
```

Please apply this where needed.

File: pkgs/development/tools/hindsight/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/hindsight/default.nix:1]:
```suggestion
{ lib, fetchFromGitHub }:
```

You should never add pkgs to inputs.

File: pkgs/development/tools/hindsight/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/hindsight/default.nix:21]:
```suggestion
  nativeBuildInputs = [ nodejs ]
    ++ (with nodePackages; [
    esinstall
    sqlite-view
    terser
  ]);
```

File: pkgs/development/tools/hindsight/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/hindsight/default.nix:42]:
```suggestion
    IFS=: for path in $NODE_PATH; do
      ln -s $path/* node_modules/
    done
```

File: pkgs/development/tools/hindsight/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/hindsight/default.nix:45]: Please use substituteInPlace or a patch to notice changes. sed fails silently.

File: pkgs/development/tools/hindsight/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/hindsight/default.nix:54]: Please use pytest if the project works with that.

File: pkgs/development/tools/hindsight/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/hindsight/default.nix:56]:
```suggestion
    if [[ -d $out/${python.sitePackages}/pyhindsight/static/web_modules ]]; then
      echo "Javascript was not installed successfully."
      exit 1
    fi
```

File: pkgs/development/tools/hindsight/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/hindsight/default.nix:63]:
```suggestion
```

set by buildPythonApplication

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Please follow the contributing guide more closely.

File: pkgs/development/tools/hindsight/default.nix
Comment (ryneeverett) [reply pkgs/development/tools/hindsight/default.nix:45]: Is there a way to replace the whole line if there's a match with substituteInPlace?

**Comment by ryneeverett** [comment]
> Please follow the contributing guide more closely.

What document are you referring to?

File: pkgs/development/node-packages/composition.nix
Comment (ryneeverett) [reply pkgs/development/node-packages/composition.nix:9]: If I understand the situation correctly it may make sense to hold off on merging anything touching nodePackages until the node2nix update is complete. I would then think I could re-run node2nix without producing these diffs.

**Comment by SuperSandro2000** [comment]
> > Please follow the contributing guide more closely.
> 
> What document are you referring to?

https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md and https://github.com/NixOS/nixpkgs/tree/master/doc/contributing

File: pkgs/development/tools/hindsight/default.nix
Comment (SuperSandro2000) [reply pkgs/development/tools/hindsight/default.nix:47]: You can input the entire line but not by regex. Using sed when requiring regex is okay.

File: pkgs/development/node-packages/composition.nix
Comment (SuperSandro2000) [reply pkgs/development/node-packages/composition.nix:9]: Yeah, kinda. It is not only this change but stdenv.lib to lib migration which also broke some things.

File: pkgs/development/node-packages/node-env.nix
Comment (jonringer) [inline pkgs/development/node-packages/node-env.nix:3]: utillinux is the alias, please revert

File: pkgs/development/python-modules/puremagic/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/puremagic/default.nix:4]: Using `pythonXPackages` isn't allowed from `python-modules` as this will mix interpreter versions for a given package. For example, if you were to do `python37Packages.<mypkg>`, but the expression pulled from `python3[8]Packages`. Python will be mis-importing several libraries, and this will likely cause issues at runtime.

Please refer to [Python dependency handling](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#handling-dependencies) on the best practices for handling dependencies within `python-modules`. Namely, you should be importing each python package individually, and not refer to another python package set (E.g. `python3Packages`).

If you're new to nixpkgs, you're free to view some of my videos on nixpkgs and python:
- [Python packaging using nixpkgs](https://www.youtube.com/watch?v=jXd-hkP4xnU)
- [Adding a package to nixpkgs unstable](https://www.youtube.com/watch?v=fvj8H5yUKu8)

File: pkgs/development/python-modules/xlsx2html/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/xlsx2html/default.nix:4]: Using `pythonXPackages` isn't allowed from `python-modules` as this will mix interpreter versions for a given package. For example, if you were to do `python37Packages.<mypkg>`, but the expression pulled from `python3[8]Packages`. Python will be mis-importing several libraries, and this will likely cause issues at runtime.

Please refer to [Python dependency handling](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#handling-dependencies) on the best practices for handling dependencies within `python-modules`. Namely, you should be importing each python package individually, and not refer to another python package set (E.g. `python3Packages`).

If you're new to nixpkgs, you're free to view some of my videos on nixpkgs and python:
- [Python packaging using nixpkgs](https://www.youtube.com/watch?v=jXd-hkP4xnU)
- [Adding a package to nixpkgs unstable](https://www.youtube.com/watch?v=fvj8H5yUKu8)

File: pkgs/top-level/all-packages.nix
Comment (jonringer) [inline pkgs/top-level/all-packages.nix:12179]:
```suggestion
  hindsight = callPackage ../development/tools/hindsight { };
```

File: pkgs/development/tools/hindsight/default.nix
Comment (jonringer) [inline pkgs/development/tools/hindsight/default.nix:5]:
```suggestion
{ lib, python3, fetchFromGitHub, nodejs, nodePackages }:

with python3.pkgs;

buildPythonApplication rec {
```

easier to track what is in the namespace if you only import other top-level attrs

File: pkgs/development/tools/hindsight/default.nix
Comment (ryneeverett) [reply pkgs/development/tools/hindsight/default.nix]: FYI, there are a couple examples of passing pkgs as inputs in the python language doc you linked: https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#handling-dependencies

**Comment by SuperSandro2000** [comment]
I am going to block this until #110394 is merged.

**Comment by raboof** [comment]
> I am going to block this until #110394 is merged.

#110545 was just merged instead

**Comment by SuperSandro2000** [comment]
Closing due to inactivity from author.

**Pull request closed without merging**
